### PR TITLE
chore: move swappers into packages

### LINKED ISSUE #5739: Move swappers into a package that excludes Redux
### Overview

The crux of this ticket is effectively moving the `lib/swapper` directory into it's own package that does not have access to Redux.

This directory should already be decoupled, the intent of doing this is to ensure that developers do not accidentally re-introduce dependencies on Redux in the future.

### References and additional details

This should mainly be a plumbing task, as most of the work to remove dependencies should already be complete.

If there are remaining, uncaught dependencies we'll want to pass the needed data into each function as args instead.

### Acceptance Criteria

A new package exists inside the `web` monorepo, the consumption of which replaces all usages of functions inside `lib/swapper`.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Move the swappers into `@shapefshiftoss/swapper` with the goal of maintaining a clear boundary between API abstractions and react.

### What this includes

* New package `@shapeshiftoss/utils` which contains many of the shared functions between web and swappers
* Moving of files to `@shapeshiftoss/swapper` and `@shapeshiftoss/utils`
* Moving of functions to /packages to `@shapeshiftoss/swapper` and `@shapeshiftoss/utils`
* Updated imports in `web` to point to new locations
* Refactoring of swappers and tests to include the dependency injection of deps from `web`

### What this excludes

There are many opportunities to clean things up made much clearer in this PR, which are deliberately not actioned here due to the obscene surface area of the changes. We should follow up with some analysis to generate follow-up tasks once this PR is merged.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #5739

High risk.

> What protocols, transaction types or contract interactions might be affected by this PR?

* All protocols supported for swapping
* All thorchain transaction types apart from sends, including swaps LP staking defi etc
* rFOX arbitrum bridge

## Testing

This PR will require some serious testing to ensure we dont introduce a production issue:
- [ ] All trade types of LIFI
    - [ ] swaps
    - [ ] cross chain
    - [ ] multi-hop
    - [ ] affiliate fees
    - [ ] slippage
- [ ] CoW swap trades
    - [ ] swaps
    - [ ] affiliate fees
    - [ ] slippage
- [ ] 0x trades
    - [ ] swaps
    - [ ] affiliate fees
    - [ ] slippage
- [ ] All trade types of THORChain
    - [ ] swaps on UTXO, Cosmos, EVM
    - [ ] streaming swaps on UTXO, Cosmos, EVM
    - [ ] affiliate fees
    - [ ] slippage
- [ ] rFOX Arbitrum bridge
- [ ] LP deposits
- [ ] LP withdrawals
- [ ] Lending
- [ ] THORChain savers

### Engineering

Pay particular attention to:

- [ ] L1 gas estimation in lifi was migrated from ethers to viem
- [ ] Cowswap `react-app-rewired` changes to silence a webpack warning which should be benign

## Screenshots (if applicable)

THORChain swap

image

THORChain streaming swap

image
image

0x trade

image

LIFI trade

image

CoW swap trade

image

rFOX bridge

image

rFOX stake

image

LP withdrawal

image

LP consolidation and deposit

image
image

### CODE REVIEW

File: src/state/slices/tradeQuoteSlice/helpers.ts
Comment (woodenfurniture) [inline src/state/slices/tradeQuoteSlice/helpers.ts:154]: moved to `@shapeshiftoss/swapper`

File: src/state/slices/assetsSlice/assetsSlice.ts
Comment (woodenfurniture) [inline src/state/slices/assetsSlice/assetsSlice.ts:69]: moved to `@shapeshiftoss/utils`

File: packages/swapper/src/swappers/CowSwapper/utils/helpers/helpers.ts
Comment (woodenfurniture) [inline packages/swapper/src/swappers/CowSwapper/utils/helpers/helpers.ts:274]: new cowswap sdk uses a number for `slippageBips`

File: package.json
Comment (woodenfurniture) [inline package.json:172]: pretty-ms 8 is a pure ESM module which fails when importing from web

File: src/lib/utils/index.ts
Comment (woodenfurniture) [inline src/lib/utils/index.ts:234]: moved to @shapeshiftoss/utils

File: src/lib/utils/index.ts
Comment (woodenfurniture) [inline src/lib/utils/index.ts:219]: moved to @shapeshiftoss/utils

File: src/lib/utils/index.ts
Comment (woodenfurniture) [inline src/lib/utils/index.ts:187]: moved to @shapeshiftoss/swapper

File: src/lib/utils/evm/index.ts
Comment (woodenfurniture) [inline src/lib/utils/evm/index.ts:287]: all of these moved to @shapeshiftoss/utils

File: src/lib/types.ts
Comment (woodenfurniture) [inline src/lib/types.ts:7]: moved to @shapshiftoss/types

File: src/constants/constants.ts
Comment (woodenfurniture) [inline src/constants/constants.ts:13]: moved to @shapeshiftoss/swapper constants

File: scripts/generateAssetData/ethereum/uniswapV2Pools.ts
Comment (woodenfurniture) [inline scripts/generateAssetData/ethereum/uniswapV2Pools.ts:61]: now using the one in @shapeshiftoss/utils

File: react-app-rewired/index.ts
Comment (woodenfurniture) [inline react-app-rewired/index.ts:363]: Tried bumping cowswap version to latest - didnt help but broke tests, so reverted to current version.

File: packages/swapper/src/types.ts
Comment (woodenfurniture) [inline packages/swapper/src/types.ts:28]: an example of something I want to clean up after this is merged

File: packages/swapper/src/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
Comment (NeOMakinG) [inline packages/swapper/src/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts:119]: suggestion: this interface doesn't look very sane and easy to consume, I would expect to have a more common interface such as 

`getL1Quote(input, config)` where config includes dependencies and any other configurations needed to get the l1 quote such as `streamingInterval` and `tradeType`

Same comment for any other `getXQuote` function with dependencies 

I shouldn't block this PR but it might be worth tackling it in a follow up to simplify it from a DX point of view

File: packages/swapper/src/types.ts
Comment (NeOMakinG) [reply packages/swapper/src/types.ts:28]: Agree it sucks as it's not react specific anymore, but this is just about naming anyway

File: packages/swapper/src/utils.ts
Comment (NeOMakinG) [inline packages/swapper/src/utils.ts:193]: nitpick: This file looks like my drawer where I throw all the things I don't wan't to see on my desk, we might want to separate the types and utils into smaller chunks

File: packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts
Comment (NeOMakinG) [inline packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:11]: question: Do we really need to import this `dist` folder? I would expect to be able to import the source file, it makes it 100% needed to build this as a dependency in order to consume it, making the DX a bit harder

File: react-app-rewired/index.ts
Comment (NeOMakinG) [reply react-app-rewired/index.ts:363]: Weird that we didn't have this warning before, wait, did we have it? What made it out loud?

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts
Comment (NeOMakinG) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:4]: suggestion: import it from the package and not the dist, I might missed other places where we do this

File: src/pages/ThorChainLP/components/ReusableLpStatus/TransactionRow.tsx
Comment (NeOMakinG) [inline src/pages/ThorChainLP/components/ReusableLpStatus/TransactionRow.tsx:15]: suggestion: Same thing with the swapper package, we might prefer to import the source code instead of the dist file

**Review by NeOMakinG — COMMENTED** [review]
I didn't test at runtime yet, nothing blocking but a few suggestions

Having an utils package makes a lot of sense, while having a swapper package doesn't make so much sense to me, the value decrease as long as we don't use it outside of our project but as it shows a few potential improvements and doesn't feel so risky, I'm still happy with it!

**Comment by NeOMakinG** [comment]
# Runtime tests
- [x] All trade types of LIFI
    - [x] swaps
    - [x] cross chain
    - [x] multi-hop 
    - [x] affiliate fees
    - [x] slippage
- [x] CoW swap trades
    - [x] swaps
    - [x] affiliate fees
    - [x] slippage
- [x] 0x trades
    - [x] swaps
    - [x] affiliate fees
    - [x] slippage
- [ ] All trade types of THORChain
    - [x] swaps on UTXO, Cosmos, EVM
    - [ ] streaming swaps on UTXO, Cosmos, EVM
    - [x] affiliate fees
    - [x] slippage
    - [x] longtail
- [x] rFOX Arbitrum bridge
- [x] LP deposits
- [x] LP withdrawals
- [ ] Lending (can't test)
- [x] THORChain savers
## Li.Fi Allbridge
ETH.ARB => AVAX, seems good and
image

Multihop
image

## Thorchain longtail
image

## Cowswap quote
image

## 0x
image

## LP deposit
image

## LP Withdraw
image

## Thorchain savers
image

## rFOX bridge
image

## Thorchain AVAX => BTC
image

## Thorchain AVAX => BCH
image
Inbound: https://avascan.info/blockchain/c/tx/0xA2AC2BE10AC3F3E1368975F7B20769D8BAA9E776E51E852F90F5D88E0AFCB3B8
Outbound: https://viewblock.io/thorchain/tx/A2AC2BE10AC3F3E1368975F7B20769D8BAA9E776E51E852F90F5D88E0AFCB3B8

I'll leave streaming swaps for the ops but memo looks fine!

LGTM at runtime!

**Review by NeOMakinG — APPROVED** [review]
Runtime tests actually looks good to me, comments are not blocking

File: packages/swapper/src/utils.ts
Comment (0xApotheosis) [reply packages/swapper/src/utils.ts:193]: We all have one of those drawers 😂

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts
Comment (0xApotheosis) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:4]: +1, probably an auto import issue.

**Review by 0xApotheosis — APPROVED** [review]
Ser, what a monster PR.

It was lot to review, and so I'm sure 1000x more to author. Purposely kept any nitpick comments to myself, lets get this thing in.

In addition to Neo's excellent testing, I also tested:

✅ Swap approvals
✅ LiFi swaps
✅ 0x swaps
✅ CoW swaps
✅ THORChain swaps

Plus a general app smoke test.

File: src/pages/ThorChainLP/components/ReusableLpStatus/TransactionRow.tsx
Comment (woodenfurniture) [reply src/pages/ThorChainLP/components/ReusableLpStatus/TransactionRow.tsx:15]: I dont think we ever want to import from packages via `/src` as that'd be treating them as a directory rather than a package which defeats the purpose of having a package.

Better yet, we should move these thorchain helpers into a thorchain package so we arent forced to import deeply from the bowels of swapper

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts
Comment (woodenfurniture) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:4]: Agree, see comment https://github.com/shapeshift/web/pull/7249#discussion_r1663454578

**Comment by woodenfurniture** [comment]
@NeOMakinG 
> Weird that we didn't have this warning before, wait, did we have it? What made it out loud?

Its due to the different compiler options of packages vs web, where webpack is unable to interpret the import as an expression unlike the compiled output of web which uses different configuration

File: packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ArbitrumBridgeSwapper/utils/fetchArbitrumBridgeSwap.ts:11]: Yes for now, see https://github.com/shapeshift/web/pull/7249#discussion_r1663454578 where we should be moving things into more ergonomic locations to avoid importing from dist

File: packages/swapper/src/utils.ts
Comment (woodenfurniture) [reply packages/swapper/src/utils.ts:193]: Agree, another one of those things which is obvious now its in a sane location - lets clean up in follow up 🙏

File: packages/swapper/src/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ThorchainSwapper/getThorTradeQuote/getTradeQuote.ts:119]: Also agree on this one, but its important we minimise diff in this mega PR so lets do in follow-up

File: react-app-rewired/index.ts
Comment (woodenfurniture) [reply react-app-rewired/index.ts:363]: Its due to the different compiler options of packages vs web, where webpack is unable to interpret the import as an expression unlike the compiled output of web which uses different configuration

File: packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts
Comment (woodenfurniture) [reply packages/swapper/src/swappers/ArbitrumBridgeSwapper/getTradeQuote/getTradeQuote.ts:55]: For the chain adapter getters, they are tied to runtime react config via feature flags which are deeply rooted in the react context. The chain adapter manager is also deeply ties to react context, so this mechanism allows us to access chain adapters without any concept of react

**Pull request merged**
